cpp/ql/src/Security/CWE/CWE-367/TOCTOUFilesystemRace.ql

```diff
@@ -89,44 +89,66 @@ predicate referenceTo(Expr source, Expr use) {
   )
 }
 
+pragma[noinline]
+predicate statCallWithPointer(Expr checkPath, Expr call, Expr e, Variable v) {
+  call = stat(checkPath, e) and
+  e.getAChild*().(VariableAccess).getTarget() = v
+}
+
+predicate checksPath(Expr check, Expr checkPath) {
+  // either:
+  // an access check
+  check = accessCheck(checkPath)
+  or
+  // a stat
+  check = stat(checkPath, _)
+  or
+  // access to a member variable on the stat buf
+  // (morally, this should be a use-use pair, but it seems unlikely
+  // that this variable will get reused in practice)
+  exists(Expr call, Expr e, Variable v |
+    statCallWithPointer(checkPath, call, e, v) and
+    check.(VariableAccess).getTarget() = v and
+    not e.getAChild*() = check // the call that writes to the pointer is not where the pointer is checked.
+  )
+}
+
+predicate checkUse(Expr check, Expr checkPath, FunctionCall use, Expr usePath) {
+  // `check` looks like a check on a filename
+  checksPath(check, checkPath) and
+  // `op` looks like an operation on a filename
+  use = filenameOperation(usePath)
+  or
+  // another filename operation (null pointers can indicate errors)
+  check = filenameOperation(checkPath) and
+  // `op` looks like a sensitive operation on a filename
+  use = sensitiveFilenameOperation(usePath)
+}
+
+pragma[noinline]
+Expr getACheckedPath(Expr check, SsaDefinition def, StackVariable v) {
+  checkUse(check, result, _, _) and
```

Comment:
Why the repeated call to `checkuse` here? Its already called on line 142. 'result' here would presumably be the same as `checkPath` from the caller point, so why not just pass the `checkPath` to the function and avoid the repeated call to `checkuse`? 

Same question for line 136.

Reply:
Those repeated calls is what I referred to when I wrote that we "manually add context into those predicates". We want to ensure that we don't join `def.getAUse(v)` with itself before we restrict the set of tuples by joining with `checkUse`.

With that said, I think I've pushed another 173cefd7e4e252219b8ee663fdc6a2f6640bd05e that hopefully makes this better: Now all the columns are included in both of those predicates, and the select clause is just a giant join-on-first-6-columns now.